### ISSUE TITLE
REQ_WCAG_1.2.1 Video-Only 

### ISSUE THREAD

**Issue by ecureuill** [solex2006/SELIProject#9]
# GUIDELINE
## Principle
> PERCEIVABLE: Information and user interface components must be presentable to users in ways they can perceive. This means that users must be able to perceive the information being presented (it can't be invisible to all of their senses) Provide text alternatives for non-text content. Provide captions and other alternatives for multimedia. Create content that can be presented in different ways, including by assistive technologies, without losing meaning. Make it easier for users to see and hear content.

## Level of Conformance 
> **A** (the minimum level of conformance), the Web page satisfies all the Level A Success Criteria, or a conforming alternate version is provided

## Referenced Requirements
Reference requirements that should be meet together
- [ ] #8 

## General description
Either an alternative for time-based media **or** an audio track is provided that presents equivalent information for prerecorded video-only content.

## Intent
> Alternatives for time-based media that are text based make information accessible because text can be rendered through any sensory modality (for example, visual, auditory or tactile) to match the needs of the user. In the future, text could also be translated into symbols, sign language or simpler forms of the language (future).

The purpose of the transcript is to provide an equivalent to what is presented visually. For prerecorded video content, authors have the option to provide an audio track. 

The purpose of the audio alternative is to be an equivalent to the video. This makes it possible for users with and without vision impairment to review content simultaneously. The approach can also make it easier for those with cognitive, language and learning disabilities to understand the content because it would provide parallel presentation.

## Examples

1. An animation that illustrates how a car engine works
An animation shows how a car engine works. There is no audio and the animation is part of a tutorial that describes how an engine works. Since the text of the tutorial already provides a full explanation, the media is an alternative for text and the text alternative includes only a brief description of the animation and refers to the tutorial text for more information.
2. A video-only file with an audio track
A silent movie includes an audio track which includes a description of the action in the video.

## Sufficient Techniques

> Sufficient techniques are reliable ways to meet the success criteria.
> - From an author's perspective: If you use the sufficient techniques for a given criterion correctly and it is accessibility-supported for your users, you can be confident that you met the success criterion.
> - From an evaluator's perspective: If web content implements the sufficient techniques for a given criterion correctly and it is accessibility-supported for the content's users, it conforms to that success criterion. (The converse is not true; if content does not implement these sufficient techniques, it does not necessarily fail the success criteria, as explained in Testing Techniques below.)
> There may be other ways to meet success criteria besides the sufficient techniques in W3C's Techniques for WCAG document, as Other Techniques below. (See also Techniques are Informative above.)

- G158 OR G166

### [G158: Providing an alternative for time-based media for audio-only content](https://www.w3.org/WAI/WCAG21/Techniques/general/G158)
>In order to present the same information in accessible form, this technique involves creating a document that tells the same story and presents the same information as the prerecorded video-only content. In this technique, the document serves as a long description for the content and includes all of the important information as well as descriptions of scenery, actions, expressions, etc. that are part of the presentation

**Example**
An animation shows how to assemble a woodworking project. There is no audio, but the animation includes a series of numbers to represent each step in the process as well as arrows and picture-in-picture highlights illustrating how the assembly is completed. It also includes short outtake animations illustrating what will happen if assembly is done incorrectly. A text alternative that identifies the video-only content reads, "Breadbox assembly video (text description follows)," and the text description of the video includes a full text description of each step in the video.

#### Test Procedure

1. View the video-only content while referring to the alternative for time-based media.
2. Check that the dialogue in the transcript matches the dialogue and information presented in the video-only presentation.
3. If the video includes multiple people or characters, check that the transcript identifies which person or character is associated with each action described.
4. Check that at least **one of the following is true**:
- The transcript itself can be programmatically determined from the text alternative for the video-only content
- The transcript is referred to from the programmatically determined text alternative for the video-only content
5. If the alternate version(s) are on a separate page, check for the availability of link(s) to allow the user to get to the other versions.

**Expected Results**

All of the above checks are true.

### [G166: Providing audio that describes the important video content and describing it as such](https://www.w3.org/WAI/WCAG21/Techniques/general/G166)
>Video-only content is inaccessible to people who are blind and to some who have low vision. Therefore, it is important for them to have an audio alternative. One way of doing this is to provide an audio track describing the information in the video. The audio should be a common audio format used on the internet, such as MP3.

**Example**
A Web page has a link to a video-only presentation of a spaceship landing on Mars. The link to the video is a picture of a spaceship. Near the video is a link to an audio file of a person describing the video. This would look something like the following code example in HTML.
```
 <a href="../video/marslanding.mp4"><img src="../images/spaceship.jpg" 
                alt="Mars landing, video-only" width="193" height="255"/></a>
                <br />
                <a href="Mars_landing_audio.mp3">Audio description of "Mars Landing"</a>
```

#### Test Procedure

For a Web page that contains video-only content:
1. Check that there is link to an audio alternative which describes the contents of the video immediately before or after the video-only content.

**Expected Results**

Check #1 is true

## Advisory Techniques

> Advisory techniques are suggested ways to improve accessibility. They are often very helpful to some users, and may be the only way that some users can access some types of content.
> Advisory techniques are not designated as sufficient techniques for various reasons such as:
> - they may not be sufficient to meet the full requirements of the success criteria;
> - they may be based on technology that is not yet stable;
> - they may not be accessibility supported in many cases (for example, assistive technologies do not work with them yet);
> - they may not be testable;
> - in some circumstances they may not be applicable or practical, and may even decrease accessibility for some users while increasing it for others;
> - they may not address the success criterion itself, and instead provide related accessibility benefits.
> Authors are encouraged to apply all of the techniques where appropriate to best address the widest range of users' needs.

### [H96:  Using the track element to provide audio descriptions](https://www.w3.org/WAI/WCAG21/Techniques/html/H96.html)
> The objective of this technique is to use the HTML5 track element to specify a descriptions timed text track for a video element. Audio description timed text tracks contain textual descriptions of the video component of the media resource, intended for audio synthesis when the visual component is obscured, unavailable, or not usable. The user agent makes the cues available to the user in a non-visual fashion, for instance, by synthesizing them into speech.

> :exclamation: As of February 2019 when this Advisory technique was last reviewed by the Working Group, there is no native support in user agents for this technique. However, support is available via JavaScript polyfills.

**Example**
A video element for a video in the English language. The audio descriptions are provided in the WebVTT format.
```
<video poster="myvideo.png" controls>
  <source src="myvideo.mp4" srclang="en" type="video/mp4">
  <track src="myvideo_en.vtt" kind="descriptions" srclang="en" label="English">
</video>
```

#### Test Procedure

For each video element used to play a video:
1. Check that the video contains a track element of kind descriptions in the language of the video.

**Expected Results**

Check #1 is true

# SELI CONTEXT
> How to implement this requirement on SELI Platform?

```
<div id="id_video_box">
   <h3>Video title, video-only</h3>
   <p  id="id_short_desc"><!-- SHORT DESCRIPTION --></p>
   <h3><button aria-expanded="false" aria-controls="id_req_acc">Available accessibility resources</button></h3>
   <ul id="id_req_acc">
  	<li>Audio description</li>
 	<li>Transcriptions</li>
   </ul>
   <video id=video poster="poster.png" autoplay="false" aria-labelby= "id_short_desc" aria-describedby="id_long_desc">
       <source id='mp4' src="trailer.mp4" type='video/mp4'/>
       <!-- Track to be used for AUDIO-DESCRIPTION as text (VTT). It will be Synthesized as audio -->
       <track kind="descriptions" label="English[desc]" src="audiodesc_de.vtt" srclang="en"></track>
       <p>Your user agent does not support the HTML5 Video element.</p>
   </video>
</div>

<button aria-expanded="false" aria-controls="id_long_desc">Video Transcription</button>
<div><p  id="id_long_desc"><!-- TRANSCRIPTIONS --></p></div>

<div> <!-- AUDIO-DESCRIPTION as audio file -->
  <p id="id_audio_short_desc"><!-- Audio description of Video Title --></p>
  <audio id=audio autoplay="false" aria-labelby="id_audio_short_desc">
    <source src="audio.mp3" type='audio/mpeg'/>
  </audio>
</div>
```
## Tasks
List the tasks required to accomplish the requirement (not exhaustive)

- [ ] **DEVELOPERS** MUST include video-only contents with it _text alternative_ AND/OR _audio alternative_
> All contents not created by users (e.g. SELI "about" section text, buttons, fields) that STUDENT could consume 

- When **TEACHER** uploads a **video content**
  - [ ] **SYSTEM** MUST provide a checkbox to **TEACHER** marks if it is a video-only.
- IF **TEACHER** uploads a **video-only content**, i.e. checks the video-only checkbox 
  - [ ] **SYSTEM** MUST provide:  
    - an option to **TEACHER** write/paste a text transcription; OR 
    - an option to **TEACHER** uploads an audio file (:exclamation: or a .vtt file), as an an audio-description.

- [ ] **DEVELOPERS** MUST render teacher's content with it respective text alternative, if teacher configured one.

# :notebook_with_decorative_cover:   References

- [How to Meet WCAG (Quick Reference)](https://www.w3.org/WAI/WCAG21/quickref/#audio-only-and-video-only-prerecorded)
- [Understanding Techniques for WCAG Success Criteria](https://www.w3.org/WAI/WCAG21/Understanding/understanding-techniques)
- [aria-describedby for longer text alternative](https://www.w3.org/WAI/PF/HTML/wiki/Media_Alt_Technologies#2:_Use_.40aria-describedby_for_the_longer_text_alternative)
- [Transcript Examples](https://www.opaltranscriptionservices.com/transcript-example/)

# GLOSSARY 
:warning: The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119/8174](http://www.rfc-editor.org/info/rfc8174).
1. **MUST** This word, or the terms "REQUIRED" or "SHALL", mean that the definition is an absolute requirement of the specification.

2. **MUST NOT** This phrase, or the phrase "SHALL NOT", mean that the definition is an absolute prohibition of the specification.

3. **SHOULD** This word, or the adjective "RECOMMENDED", mean that there may exist valid reasons in particular circumstances to ignore a particular item, but the full implications must be understood and carefully weighed before choosing a different course.

4. **SHOULD NOT** This phrase, or the phrase "NOT RECOMMENDED" mean that there may exist valid reasons in particular circumstances when the particular behavior is acceptable or even useful, but the full implications should be understood and the case carefully weighed before implementing any behavior described with this label.

5. **MAY**   This word, or the adjective "OPTIONAL", mean that an item is truly optional. An implementation which does not include a particular option MUST be prepared to interoperate with another implementation which does include the option, though perhaps with reduced functionality. In the same vein an implementation which does include a particular option MUST be prepared to interoperate with another implementation which does not include the option (except, of course, for the feature the option provides.)

**Comment by ecureuill** [comment]
For the video-only content be accessible (level A), teacher just need to include one alternative content (transcription or audio-description). So system just need to provide one option to the teacher configure the accessibility resource.

But **I think** that teachers could benefit from a more flexible approach if system provides one or more options, e.g. a teacher may find it easier to produce a transcript for video A and easier to produce an audio for video B.

@solex2006 could you check with others which approach to follow? This is more a business rule decision.
